ecosystem/sep-XDRJSON.md

```diff
@@ -0,0 +1,1058 @@
+## Preamble
+
+```
+SEP: To Be Assigned
+Title: XDR-JSON
+Author: Leigh McCulloch
+Track: Standard
+Status: Draft
+Created: 2025-04-16
+Updated: 2025-04-16
+Version: 0.1.0
+Discussion: https://github.com/stellar/stellar-protocol/discussions/1702
+```
+
+## Simple Summary
+
+This proposal defines XDR-JSON, a standard mapping between Stellar's XDR
+(External Data Representation) structures and a JSON representation.
+
+## Dependencies
+
+None.
+
+## Motivation
+
+Stellar's protocol data is defined in XDR, a compact binary format that is not
+human-readable. In most APIs when the XDR must be shared as text it is base64
+encoded, which is also not human-readable or easily mutated. Developer tooling
+needs to represent this data in a more accessible format either so it can be
+read or modified.
+
+## Abstract
+
+This proposal defines how a subset of the XDR data types defined in [RFC 4506]
+is mapped to a JSON representation. It covers primitive types (integers,
+booleans), complex types (structs, unions, enums), and special Stellar-specific
+types (AccountID, Asset, etc.). The specification ensures that XDR data can be
+consistently serialized to JSON and deserialized back to XDR without data loss,
+while providing human readability, specifically developers.
+
+## Specification
+
+### XDR Data Types
+
+#### Integer (32-bit)
+
+The XDR 32-bit signed integer data type ([RFC 4506 Section 4.1]) maps to JSON
+numbers.
+
+For example:
+
+XDR Definition:
+
+```xdr
+int identifier;
+```
+
+XDR Binary:
+
+```b
+00000000: 7fff ffff                                ....
+```
+
+XDR Binary Base64 Encoded:
+
+```base64
+f////w==
+```
+
+JSON:
+
+```json
+2147483647
+```
+
+#### Unsigned Integer (32-bit)
+
+The XDR 32-bit unsigned integer data type ([RFC 4506 Section 4.2]) maps to JSON
+numbers.
+
+For example:
+
+XDR Definition:
+
+```xdr
+unsigned int identifier;
+```
+
+XDR Binary:
+
+```b
+00000000: ffff ffff                                ....
+```
+
+XDR Binary Base64 Encoded:
+
+```base64
+/////w==
+```
+
+JSON:
+
+```json
+4294967295
+```
+
+#### Hyper Integer (64-bit)
+
+The XDR 64-bit signed integer data type ([RFC 4506 Section 4.5]) maps to JSON
+numbers.
+
+For example:
+
+XDR Definition:
+
+```xdr
+hyper identifier;
+```
+
+XDR Binary:
+
+```b
+00000000: 7fff ffff ffff ffff                      ........
+```
+
+XDR Binary Base64 Encoded:
+
+```base64
+f/////////8=
+```
+
+JSON:
+
+```json
+9223372036854775807
+```
+
+_Note: JavaScript runtime implementations of JSON do not support numbers that
+require more than 53-bits. JavaScript applications should use a JSON decoder
+supporting 64-bit numbers in JSON numbrs. JSON decoders in other languages do
+not typically have this constraint and support numbers up to 64-bits._
```

Comment:
Unfortunately, I don't think this is an appropriate constraint to place on the most popular SDK language :/ negating the ability to use the built-in [`JSON.parse`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse) and requiring a separate parsing library leads to all of the classic JS world problems: environment compatibility, increased bundle sizes, etc. (Though they [do exist](https://stackoverflow.com/a/18755261).)

It also counters conventions in Horizon and RPC which encode all >= 64-bit values as strings. (Separately, I can't find the discussion anymore but I do think 32-bit integers should stay `number`.)

Reply:
Yup, that's why we have this open issue to change it, and also change the presentation of the 128 and 256 bit integer types too:
- https://github.com/stellar/xdrgen/issues/181#issuecomment-2819617408

This PR captures the existing XDR-JSON format which has this wart unfortunately.